src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,746 @@
+ *
+ *     private static final IntFunction<Integer> FIB =
+ *             StableValue.intFunction(MAX_SIZE_INT, Fibonacci::fib);
+ *
+ *     public static int fib(int n) {
+ *         return n < 2
+ *                 ? n
+ *                 : FIB.apply(n - 1) + FIB.apply(n - 2);
+ *     }
+ *
+ * }
+ *}
+ * Both {@code FIB} and {@code Fibonacci::fib} recurses into each other. Because the
+ * stable int function {@code FIB} caches intermediate results, the initial
+ * computational complexity is reduced from exponential to linear compared to a
+ * traditional non-caching recursive fibonacci method. Once computed, the VM is free to
+ * constant-fold expressions like {@code Fibonacci.fib(10)}.
+ * <p>
+ * The fibonacci example above is a dependency graph with no circular dependencies (i.e.,
+ * it is a dependency tree):
+ *{@snippet lang=text :
+ *
+ *              ___________fib(5)____________
+ *             /                             \
+ *       ____fib(4)____                  ____fib(3)____
+ *      /              \                /              \
+ *    fib(3)          fib(2)          fib(2)          fib(1)
+ *   /     \         /     \         /     \
+ * fib(2) fib(1)   fib(1) fib(0)   fib(1) fib(0)
+ *}
+ *
+ * If there are circular dependencies in a dependency graph, a stable value will
+ * eventually throw a {@linkplain StackOverflowError} upon referencing elements in
+ * a circularity.
+ *
+ * <h2 id="thread-safety">Thread Safety</h2>
+ * The content of a stable value is guaranteed to be set at most once. If competing
+ * threads are racing to set a stable value, only one update succeeds, while other updates
+ * are blocked until the stable value becomes set.
+ * <p>
+ * The at-most-once write operation on a stable value that succeeds
+ * (e.g. {@linkplain #trySet(Object) trySet()})
+ * {@linkplain java.util.concurrent##MemoryVisibility <em>happens-before</em>}
+ * any subsequent read operation (e.g. {@linkplain #orElseThrow()}) that is successful.
+ * A successful write operation can be either:
+ * <ul>
+ *     <li>a {@link #trySet(Object)} that returns {@code true},</li>
+ *     <li>a {@link #setOrThrow(Object)} that does not throw, or</li>
+ *     <li>an {@link #orElseSet(Supplier)} that successfully runs the supplier</li>
+ * </ul>
+ * A successful read operation can be either:
+ * <ul>
+ *     <li>a {@link #orElseThrow()} that does not throw,</li>
+ *     <li>a {@link #orElse(Object) orElse(other)} that does not return the {@code other} value, or</li>
+ *     <li>an {@link #orElseSet(Supplier)} that does not {@code throw}</li>
+ * </ul>
+ * <p>
+ * The method {@linkplain StableValue#orElseSet(Supplier) orElseSet()} guarantees that
+ * the provided {@linkplain Supplier} is invoked successfully at most once even under
+ * race. Since stable functions and stable collections are built on top of
+ * {@linkplain StableValue#orElseSet(Supplier) orElseSet()} they too are
+ * thread safe and guarantee at-most-once-per-input invocation.
```

Comment:
Do we guarantee that there is a total order with respect to the suppliers that fail to return. You can take a look at the specification of [`std::call_once`](https://en.cppreference.com/w/cpp/thread/call_once)

> All active calls on the same flag form a single total order consisting of zero or more exceptional calls, followed by one returning call. The end of each active call synchronizes-with the next active call in that order.
> The return from the returning call synchronizes-with the returns from all passive calls on the same flag: this means that all concurrent calls to std::call_once are guaranteed to observe any side-effects made by the active call, with no additional synchronization.

Reply:
i don't think we should promise this even though it might be the case.

Reply:
What is the benefit of not promising it?

Reply:
I tend to agree with @merykitty - what that guarantee says is that - e.g. you can't have an exceptional call to the supplier (but this holds also for `orElseSet`) _after_ the stable value has been set. I think we already guarantee this implicitly, because we say that the action in `orElseSet` is never executed if the stable value has been set -- which implies that at least a supplier terminated normally (and that must be the last one -- otherwise it would not have executed at all...)